### ISSUE TITLE
Support PSR-6/16 for caching instead of a custom interface [Request for Comments]

### ISSUE THREAD

**Issue by stadskle** [auth0/auth0-PHP#282]
Very unnecessary to have a custom interface for caching when we have a PHP-standard for this. The class should support PSR-6 and do type hinting against that, making it easier to integrate with existing frameworks.

**Comment by Nyholm** [comment]
This is fixed in the 6.x branch.

**Comment by joshcanhelp** [comment]
@stadskle - I agree and I'll see if this is possible without making a breaking change. 

Just FYI, the 6.x.x branch that @Nyholm mentioned is **not** ready for production yet and may not be released in it's current state.

**Comment by joshcanhelp** [comment]
I wanted to use this thread to discus our plans here. I realize it has been a while since this issue was raised but we're working on a major in the `7.0.0-dev` branch now and can finally tackle this in the right way. We'd love some community help on this one since the caching objects themselves will be provided by the developer. 

Caching in this library is currently used to cache JWKS files pulled from the Auth0 server in the login callback. At some point, it will also be used to cache the discovery doc from the same place. 

We would like the interface to be as universal as possible and so we are, of course, looking at PSR to help us with that. This issue mentions [PSR-6](https://www.php-fig.org/psr/psr-6/), which would work, but our needs here are very simple so it seems like [PSR-16](https://www.php-fig.org/psr/psr-16/) would do the job. It looks like Symfony provides [adapters to convert between the two](https://symfony.com/doc/current/components/cache/psr6_psr16_adapters.html) so our choice of standard here may not be critical, in which case I would lean towards PSR-16. 

In my mind, the best thing we could do here is:

1. Choose one of the standards and make sure those interfaces are implemented in the SDK
2. Provide a sample implementation in the form of a memory cache
3. Provide some guidance in the form of documentation about using other libraries to implement, say, a file system, Redis, or database cache

Does anyone have any feedback on the approach or the standard that we should choose? 

Thank you!

**Comment by Nyholm** [comment]
I would advice against psr-16. 
Psr6 is more powerful and way more common in other libraries. There are adapters between these two standards but why use an adapter layer when you can write 4 lines of code yourself. 

Just my 2 cents.

**Comment by joshcanhelp** [comment]
@Nyholm - Appreciate your input here!

> Psr6 is more powerful 

If we're only have very basic needs here, I'm not sure being powerful is an asset. 

> way more common in other libraries

That's what I'm curious about. I don't want to choose a standard that folks would have to work around (as we did originally with the custom interface).

> why use an adapter layer when you can write 4 lines of code yourself

Are you saying those adapters are that simple or something else?

**Comment by stadskle** [comment]
Quickly looking at the largest frameworks, it seems like PSR-6 is **a bit** more common and used although most seems to be supporting both (and having bridge functionality). Zend Framework and Symfony has both, Laravel does not seem to have PSR-16 as far as I can tell, but you can fix it with an external bridge.

Looking at packagist it seems like the number of packages implementing the psr interfaces are about the same 

PSR-16
<img width="647" alt="Screenshot 2019-11-01 07 48 28" src="https://user-images.githubusercontent.com/6637162/68007491-0fa31100-fc7c-11e9-8889-280ef49c7a30.png">

PSR-16
<img width="497" alt="Screenshot 2019-11-01 07 48 34" src="https://user-images.githubusercontent.com/6637162/68007535-2c3f4900-fc7c-11e9-96e0-aab2c9f9f36c.png">

Looking at the top packages when searching for cache does not show any bias either, they are either supporting both of none:

<img width="896" alt="Screenshot 2019-11-01 07 50 42" src="https://user-images.githubusercontent.com/6637162/68007585-5264e900-fc7c-11e9-8c9e-22551b0a0e9a.png">

 Bottom line, my 5 cents is that this does not matter too much. Just give us a standard, if someone is eager it would be great to support both. If not pick any and the world will be a better place regardless :)
